### DISCUSSION TITLE
Refresh Tokens from REST API with Next.js

### DISCUSSION THREAD

**Discussion by mohito12** [vercel/next.js#23459] [Help]
I'm currently making my first Next.js app and I am having issues with auth. I'm using expressjs for my API and I have sign in route that sets a refresh token as cookie and also returns an access token.

Sign-in:

```
    res.cookie(
      'zis',
      sign(payload, REFRESH_TOKEN_SECRET, {
        expiresIn: '24h',
      }),
      {
        expires: new Date(Date.now() + 24 * 3600000),
        httpOnly: true,
        secure: process.env.NODE_ENV !== 'development',
        sameSite: 'strict',
      }
    );

    res.json({
      accessToken: sign(payload, ACCESS_TOKEN_SECRET, {
        expiresIn: '15m',
      }),
    });
```

I have another route `/resfresh-token` that is used to get a new access token. All my other routes are protected via middleware `isAuth` that verifies the access token.

So far I'm able to set and get the cookie from my next js page `signin.js` and also via context I'm passing the cookie to the other protected pages by using getServerSideProps() to get a new access token and verify it.

NextJS page:

```
    export async function getServerSideProps(context) {
      const cookie = context.req.headers.cookie;
       
      const resRefreshToken = await fetch(
        `http://localhost:5000/auth/refresh-token`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            cookie: cookie,
          },
        }
      );
    
      const refreshToken = await resRefreshToken.json();
    
      var bearer = 'bearer ' + refreshToken.accessToken;
    
      const res2 = await fetch(`http://localhost:5000/api/users`, {
        headers: {
          authorization: bearer,
        },
      });
      
      const users = await res2.json();
    
      return {
        props: { users },
      };
    }
```

My question is first of all, is my approach or logic wrong and is there a way to call `/refresh-token` on every request rather than repeating the same logic on each page. I've tried `getInitialProps` in `_app.js` but I have to pass the returned access token as prop and exposed it in the DOM and I'm not sure if it's safe to.

Thank you

**Comment by Psyruss77** [comment]
Ss /start

**Reply by mohito12** [reply]
what does Ss /start means?

**Comment by umarluqman** [comment]
It's safe to pass the returned access token as props as it's stored in memory as stated here https://hasura.io/blog/best-practices-of-using-jwt-with-graphql/

You can use [Axios](https://github.com/axios/axios) to configure the base URL, injecting the auth token, and configure the refresh token logic so you are able to have that refresh token logic in every request.

**Comment by pvlcor** [comment]
Your are on the right track! Here's my approach for server-side auth, hope it'll help with your questions:

If you are using server-side cookie auth, then you don't need to pass token down to client. For server-side auth to work you basically need 3 things:
1) Write wrapper function for `getServerSideProps()`, e. g. `const getServerSideProps = withAuth(async (ctx) = {})`, to pass down to each page Boolean flag, like isLoggedIn — then in `_app.js` you can grab this boolean from `pageProps` and use it on client side not exposing access token
2)  Write another wrapper function for server-side requests — like `withAuthFetch()` — which will contain logic to perform some request with auth header and do something in case of 401 error
3) Using wrapper function from step 2 proxy all your protected APIs through individual `pages/api/...` handlers — I assume you are doing this with a middleware, also a nice solution

Example of withAuth
```
export const withAuth = (getServerSideProps) => ctx => {
    ctx.isLoggedIn = Boolean(ctx.req.cookies.token)
    return getServerSideProps(ctx)
}
```
Pseudo code example for withAuthFetch
```
export const withAuthFetch = async (ctx, {url, params} ) => {
    try {
        const { token } = req.cookies
        const response = await fetch(url, {
            headers: {
                authorization: `Bearer ${token}`,
            },
            params,
        })

        return response
    } catch (error) {
        if (error.response.status === 401) {
            delete token cookie
            redirect
        }
    }
}
```
